### DISCUSSION TITLE
setColumnOrder does not exist in TypeScript

### DISCUSSION THREAD

**Discussion by Poniav** [TanStack/table#2747] [General]
I recently moved my react table to Typescript and fixed most of problems but not this one. 

`Property 'setColumnOrder' does not exist on type 'TableInstance<object>'.ts(2339)`

`const {
    getTableProps,
    getTableBodyProps,
    headerGroups,
    rows,
    prepareRow,
    allColumns,
    setColumnOrder,
    state
  } = useTable(
    {
      columns,
      data,
      defaultColumn
    },
    useFilters,
    useSortBy,
    useColumnOrder
  )`

Version :

"react-table": "^7.5.0",
"@types/react-table": "^7.0.23",

How to fix this ? 

Anyone ? :(

**Comment by playerd0** [comment]
Having the same issue...

**Comment by cmcWebCode40** [comment]
Having same issues using typescript

**Comment by andrerpena** [comment]
Having the same issue...

**Comment by ggascoigne** [comment]
Did you find https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react-table/Readme.md (I realize that it's rather buried).  Anyway, I'd guess that your `react-table-config.d.ts` is lacking references to `UseColumnOrderInstanceProps` on `TableInstance`.

**Comment by isolationism** [comment]
A year later, this is still an issue. The entire typings definition is installed, with nothing removed, per the readme referenced above.

Other parameters — like `setHiddenColumns` for example — work just fine with no problem.

This issue is specific to somehow incorrect typing for `setColumnOrder` and it persists despite the DefinitelyTyped typings for the project being updated recently (using 7.7.10).

**Reply by tannerlinsley** [reply]
These are community maintained types for (a very soon outdated) version 7. So What @ggascoigne is suggested is your best bet. Good news is that v8 is *written* 100% in TypeScript, so there will be no more need for external definitions. What you see in the v8 types will be truth!

**Reply by isolationism** [reply]
Thank you Tanner — I appreciate all that you've done on this project. Looking forward to v8!

**Reply by ggascoigne** [reply]
You added a thumbs down to my previous solution to this problem.  But that solution works, and it's not a bug, these types use interface merging, if you don't use it, it won't work.

So what exactly is the problem?

**Reply by tannerlinsley** [reply]
@isolationism, what @ggascoigne is suggested is your best bet.

**Reply by isolationism** [reply]
Does it, ggascoigne? Or did you just reply saying "I guess" without actually verifying what you wrote works correctly?

Did you read my follow-up post below saying I have followed all the instructions and see the error despite it all only for `setColumnOrder` but not for other props like `setHiddenColumns` for which the typings declarations work?

The issue is decribed very clearly and concisely. Your proposed solution does not work, and you're labouring under the false assumption that it does.

That is the problem.

**Reply by ggascoigne** [reply]
Well I've not used setColumnOrder, but can I access it and have it compile, yes, easily.

I just added it to:
```tsx
const { getTableProps, headerGroups, getTableBodyProps, page, prepareRow, state, setColumnOrder } = instance
  ```

and it compile just fine.  It'll take me a bit longer to actually work up an example that uses it correctly.

But the important bit is that you customize the `react-table-config.ts` example.

Specifically this bit:

```ts
  export interface TableInstance<D extends Record<string, unknown> = Record<string, unknown>>
    extends UseColumnOrderInstanceProps<D>,
      UseExpandedInstanceProps<D>,
      UseFiltersInstanceProps<D>,
      UseGlobalFiltersInstanceProps<D>,
      UseGroupByInstanceProps<D>,
      UsePaginationInstanceProps<D>,
      UseRowSelectInstanceProps<D>,
      UseFlexLayoutInstanceProps<D>,
      UsePaginationInstanceProps<D>,
      UseSortByInstanceProps<D> {}
```

To access `setColumnOrder`, `TableInstance` must extend `UseColumnOrderInstanceProps`.

**Reply by tannerlinsley** [reply]
Keep it civil and don't make me lock this discussion 😅

**Reply by isolationism** [reply]
First, I apologize for my statement and tone above; I am frustrated and at fault. I beg your forgiveness.

Second, to be clear, I struggle with Typescript and its many complexities; I've worked predominantly with untyped languages in the past (JavaScript, Python). The issue is certainly my fault. I believe I have followed the instructions to the best of my understanding, but suspect some underlying tsconfig.js issue on my end.

Third, I was mistaken and the issue is not specific to the `setColumnOrder` plugin; whatever I've done wrong is affecting all of them, as I had found and implemented the typings months ago. Upon recently resuming work, the typings broke again. I was only using one plugin here after all, and it was a bad example. The rest of this post is unfortunately off-topic, so forgive that as well.
 
Fourth, I am here because I'm seeking help, and my reaction was "that didn't work for me". It wasn't intended as a slight toward the author. I've removed them anyway and ask for your understanding.

That out of the way, if I may explain a little more of my issue —

I am aware of the instructions, having followed them before locally — and it did work for me, for some time. I'm not sure what's happened between then and now, where it's not.

I did a quick search for "react-table typescript codesandbox" and forked one, updated the dependencies and added the missing types declarations; it essentially resembles my local build and produces the same error in the editor. I wonder if I might implore you to have a look.

https://codesandbox.io/s/https-github-com-tanstack-react-table-discussions-2747-jvetzj?file=/src/App.tsx

Around the creation of the table instance in App.tsx around line 70, I see this:

![image](https://user-images.githubusercontent.com/400643/159060572-9b02ee5f-ba8f-4773-8e78-fcf29736f1e4.png)

I'm confident this is in turn caused by the many errors in `react-table.config.d.ts`, each showing:

> `All declarations of 'TableOptions' must have identical type parameters.ts(2428)`
    
I'm unclear on what those conflicting type definitions might be. I have the latest `react-table` and `@types/react-table` installed. 

I implore your help. Thank you for your time and patience.

**Reply by ggascoigne** [reply]
I do understand, thanks for the comments.  We've all been there and learning typescript can seem like a learning clif rather than a gentle hill at times.  I'll look at your sandbox and get back to you.

**Reply by ggascoigne** [reply]
Off the top of my head though "All declarations of 'TableOptions' must have identical type parameters.ts(2428)" is a side effect of not setting `"skipLibCheck": true` in the tsconfig.  There is a bug that cause it, but since it's easy to work around I keep forgetting to fix it.  I don't think that that's the root of the problem, but it certainly adds to the noise.

**Reply by isolationism** [reply]
OK, good point — this was a bare minimum tsconfig from the project that is a subset of my local, which includes the parameter you suggest.  I've also added it to the CodeSandbox. 

My assumption to this point was "fewer" config options that could reproduce the issue would be better than have you examine all of mine, but I've added my entire local config to the sandbox as `tsconfig.json.local` if that helps.

**Reply by isolationism** [reply]
I have solved my own issue. It is that I am not very good at reading instructions.

I had mistakenly placed the `react-table-config.d.ts` in a directory called `types` along with other such declarations. For the sake of posterity and others who might have made my mistake:

The file must go in `@types/react-table-config.d.ts`. Not the same thing as `types/react-table-config.d.ts`, apparently.

And now I have some reading to do.

Thank you both again for your help and humble apologies.

**Reply by ggascoigne** [reply]
heh, I think that I was just getting to the same conclusion.  FYI I think that @types is a default, and you can change it using by adding `"typeRoots": ["./types"]` to your tsconfig.

**Reply by ggascoigne** [reply]
Oh, and you might find https://github.com/ggascoigne/react-table-example useful, it's a decent TypeScript and material ui example.
